### PR TITLE
feat: v4 POST with signed policy

### LINKED ISSUE #1047: [Storage] V4 Signature POST Policies
Tracking issue for work defined in companion document.

[GCS - POST with SigV4](https://docs.google.com/document/d/1zLV5kpDf1gNER-twwt0sLs4BRSnVsPEWVnkhzEbT6SM/edit)

### PR DESCRIPTION
**Pull request by jkwlui**
# DO NOT MERGE 

- until production is ready and system test is updated to execute the signed policy.

Fixes #1047 🦕

- alias `getSignedPolicy()` to `getSignedPolicyV2()`, and deprecate the former
- Refactored `getSignedPolicyV4()` to only provide `fields` and `conditions` options, removing `startsWith`, `equals`, `success*`, `acl`, `contentLengthRange` options.
- update v2 links to point to https://cloud.google.com/storage/docs/xml-api/post-object-v2

TODO: execute signed policy against production once ready

### CODE REVIEW

File: src/file.ts [hunk 1]
```diff
@@ -93,6 +94,35 @@ export interface GetSignedPolicyOptions {
...
+
+export interface GetSignedPolicyV2Callback extends GetSignedPolicyCallback {}
+
+export interface PolicyFields {
+  [key: string]: string;
+}
+
+export interface GetSignedPolicyV4Options {
+  expires: string | number | Date;
+  cname?: string;
```
Comment (frankyn) [inline src/file.ts]: bucketBoundHostname

File: src/file.ts [hunk 2]
```diff
@@ -2280,6 +2432,193 @@ class File extends ServiceObject<File> {
...
+   *   fields: {
+   *     acl: 'public-read',
+   *     'x-goog-meta-foo': 'bar',
+   *     'x-ignore-mykey': 'data'
+   *   }
+   * };
+   *
+   * file.getSignedPolicyV4(options, function(err, response) {
+   *   // response.url The request URL
+   *   // response.fields The form fields to include in the POST request.
```
Comment (frankyn) [inline src/file.ts]: Add context that this is meant to be used for a form request.

**Review by frankyn — CHANGES_REQUESTED** [review]
Small nits, overall LGTM.

**Review by frankyn — COMMENTED** [review]
Thanks @jkwlui, this LGTM.

Pending is integration tests and surface checked off.

**Review by stephenplusplus — APPROVED** [review]

File: src/file.ts [hunk 3]
```diff
@@ -2280,6 +2432,195 @@ class File extends ServiceObject<File> {
...
+   *     ['content-length-range', 0, 1024],
+   *   ],
+   *   fields: {
+   *     acl: 'public-read',
+   *     'x-goog-meta-foo': 'bar',
+   *     'x-ignore-mykey': 'data'
+   *   }
+   * };
+   *
+   * file.getSignedPolicyV4(options, function(err, response) {
```
Comment (frankyn) [inline src/file.ts]: instead of response use: `policy`.

File: src/file.ts [hunk 4]
```diff
@@ -2280,6 +2432,195 @@ class File extends ServiceObject<File> {
...
+   *
+   * @param {GetSignedPolicyV4Callback} [callback] Callback function.
+   * @returns {Promise<GetSignedPolicyV4Response>}
+   *
+   * @example
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const myBucket = storage.bucket('my-bucket');
+   *
+   * const file = myBucket.file('my-file');
```
Comment (frankyn) [inline src/file.ts]: Writing canonicals, I almost want to say you should move up `getSignedPolicyV4()` to bucket instead of file.

**Review by frankyn — CHANGES_REQUESTED** [review]

File: src/file.ts (hunk 3 above)
Comment (jkwlui) [reply src/file.ts]: Would it create ambiguity? Since the policy is the encoded JSON:
```json
{
  "conditions": ...
  "expiration": ...
}
```

Comment (frankyn) [reply src/file.ts]: oh, good point, I'll update canonical sample as well. Leave as is.

**Comment by frankyn** [comment]
@jkwlui conformance tests were updated, please make sure your tests still pass.

That's the last request, overall LGTM. Thanks for all your work!

**Comment by jkwlui** [comment]
Updated. See 63bf22c

**Review by frankyn — APPROVED** [review]
LGTM.

**Pull request merged**